R/generate.R

```diff
@@ -250,22 +251,34 @@ permute <- function(x, reps = 1, variables, ..., call = caller_env()) {
 
 permute_once <- function(x, variables, ..., call = caller_env()) {
   dots <- list(...)
+  null <- attr(x, "null")
+
+  if (!is_hypothesized(x) ||
+      !null %in% c("independence", "paired independence")) {
+     abort(
+        paste0("Permuting should be done only when doing independence ",
+               "hypothesis test. See `hypothesize()`."),
+        call = call
+     )
+  }
 
-  if (is_hypothesized(x) && (attr(x, "null") == "independence")) {
-    # for each column, determine whether it should be permuted
-    needs_permuting <- colnames(x) %in% process_variables(variables, FALSE)
-
-    # pass each to permute_column with its associated logical
-    out <- purrr::map2(x, needs_permuting, permute_column)
-    out <- tibble::new_tibble(out)
+  variables <- process_variables(variables, FALSE)
+  if (null == "independence") {
+     # for each column, determine whether it should be permuted
+     needs_permuting <- colnames(x) %in% variables
 
-    copy_attrs(out, x)
+     # pass each to permute_column with its associated logical
+     out <- purrr::map2(x, needs_permuting, permute_column)
+     out <- tibble::new_tibble(out)
   } else {
-     abort(paste0(
-      "Permuting should be done only when doing independence hypothesis test. ",
-      "See `hypothesize()`."
-    ), call = call)
+     out <- x
+     signs <- sample(c(-1, 1), nrow(x), replace = TRUE, prob = c(.5, .5))
+     out[[variables]] <- x[[variables]] * signs
```

Comment:
Do we want to surface `signs` at any point? It seems like this could be visible after `generate()`, with the `calculate()` interface carrying out a new internal step of multiplying by and selecting off the `signs` column before sending values off to `calc_impl()`.

Reply:
Where would we surface them, as a new column in the df that is the result of `calculate()`? I would think, if anywhere, they would make sense as a result of `generate()` since that's the step assigning the `signs`.

Reply:
Yes, in the results of `generate()`! I'm 50/50 on whether we ought to just go ahead and multiply by the randomly sampled `signs` in `generate()` and never surface that column or output both the response and `signs` from `generate()` and, first thing in `calculate()`, multiply the response by `signs` and then select it off.

Reply:
I feel like the only reason to surface `signs` could be a teaching reason -- to show exactly what's going on under the roof it might be good to access/visualize them.

However, I wouldn't do it at the risk of getting a confusing putout from `generate()`. I feel like it wouldn't be confusing, since columns will be clearly labeled. Do you agree?

Reply:
That sounds good—I'm with it.